---
title: "ETP Pocillopora adegenet DAPC"
author: "Mike Connelly"
date: "7/24/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

## Setup packages and working directories
```{r}
library("vcfR")
library("tidyverse")
library("adegenet")
library("ggrepel")
#
library("extrafont")
library("extrafontdb")
```

## Import sample metadata
```{r factor levels}
study_factor <- c("Vidal-Dupiol2014", "Yuan2017", "Zhou2017", "Brener-Raffali2018", "Tang2018", "Wecker2018", "Zhou2018", "Poquita-Du2019", "Zhou2019", "ConnellyEAPSI", "Chuang-Mitarai2020", "Li2020")
```
```{r}
# Import sample metadata 
samples <- read_csv("data/SampleData_MetaPocillopora_011822.csv") %>% 
  filter(Include_SNPs == TRUE) %>% 
  mutate("Sample_ID" = str_c(`Study`, `Sample`, sep = "_"))
# 
```

```{r}
#pdam_genome <- ape::read.dna("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.fasta", format = "fasta")
```
```{r}
#pdam_gff <- read.table("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.gff", sep="\t", quote="")
```

## Import VCF file
```{r imporf vcf}
vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/metapocillopora_003_filtered.vcf")
```

## Convert to GENIND object
```{r}
genind_meta <- vcfR2genind(vcf_meta)
class(genind_meta)
genind_meta
```
## Convert to other formats?
```{r}
vcfR::write.fasta()
```


## PCA
```{r}
# 
x_meta <- tab(genind_meta, freq = TRUE, NA.method = "mean")
# 
pca_meta <- dudi.pca(x_meta, center = TRUE, scale = FALSE, scannf = FALSE, nf = 41)
# ?dudi.pca()
```
```{r}
pdf("./outputs/adegenet/pca_screeplot.pdf")
dudi.pca(x_meta, center=TRUE, scale=FALSE, scannf = FALSE, nf = 4)
dev.off()
```
```{r}
s.label(pca_meta$li, xax=1, yax=2)
```
```{r}
pca_df <- pca_meta$li %>%
  rownames_to_column("Sample_ID") %>%
  mutate(Sample_ID = str_replace(`Sample_ID`, "_1$", "")) %>% 
  right_join(samples, by = "Sample_ID")
pca_df$Study <- factor(pca_df$Study, levels = study_levels, ordered = TRUE)
```
```{r PC_vars}
pc1lab <- paste("PC1: ", round(pca_meta$eig[1]/sum(pca_meta$eig)*100, 2), "%", sep = "")
pc2lab <- paste("PC2: ", round(pca_meta$eig[2]/sum(pca_meta$eig)*100, 2), "%", sep = "")
```
```{r}
# Figure
pdf("./outputs/figures/VCF_SNPs_pca.pdf", height = 7, width = 10)
pca_df %>%
  ggplot(., aes(Axis1, Axis2)) +
  geom_point( aes(fill = Study, shape = Species), size = 6, alpha = 0.5, stroke = 0.5, color = "black") + 
  # geom_text_repel(aes(label = Sample_ID), size = 2, max.overlaps = 100)
  scale_fill_manual(values = gg_color_hue(7), 
                    guide = guide_legend(override.aes = list(size = 6, shape = 21, fill = gg_color_hue(7)))) +
  scale_shape_manual(values = c(21, 24, 22),
                     guide = guide_legend(override.aes = list(size = 6, alpha = 0.5, stroke = 0.5, fill = "black"))) +
  theme_bw() +
  theme(plot.title = element_text(size = 24),
        axis.title = element_text(size = 20),
        axis.text = element_blank(),
        legend.title = element_text(size = 20),
        legend.text = element_text(size = 18),
        legend.spacing.y = unit(0.15, "in")) +
  xlab(pc1lab) + ylab(pc2lab) +
  ggtitle("PCA of phylotranscriptomic SNPs (12,049)")
dev.off()
```
```{r}
# Label and color PCA points according to library, site, etc. 
```

## DAPC
```{r}
# Find clusters
grp_meta <- find.clusters(genind_meta, max.n.clust = 20)
```
```{r}
pdf("./outputs/adegenet/pca_find-clusters.pdf", height = 5, width = 6)
grp_meta <- find.clusters(genind_meta, max.n.clust = 40, n.pca = 41)
dev.off()
```
```{r}
# Inspect output of find.clusters
grp_meta$Kstat
grp_meta$stat
grp_meta$grp
grp_meta$size
```
```{r}
# Run the analysis on the dataset using the previously inferred groups
dapc_meta <- dapc(genind_meta, grp_meta$grp)
```
```{r}
# Inspect output of dapc
dapc_meta
dapc_meta$n.pca
dapc_meta$n.da
dapc_meta$tab
dapc_meta$var
```
```{r}
scatter(dapc_meta)
```
```{r}
pdf("./outputs/adegenet/dapc_max-missing-0.25_indv-missing-0.95_depth-20_maf-0.05.pdf")
scatter(dapc_meta,
         scree.da=TRUE,  posi.da="bottomright",
         scree.pca=TRUE, posi.pca="bottomleft",
         bg="white", pch=20,  cell=0, cstar=0,
        cex=3,clab=0, leg=TRUE, txt.leg=paste("Cluster",1:12))
dev.off()
```
```{r}
scatter(dapc_meta, 1, 1, bg="white",
        scree.da=FALSE, legend=TRUE, solid=.4)
```
```{r}
scatter(dapc_meta, 2, 2, bg="white",
        scree.da=FALSE, legend=TRUE, solid=.4)
```

```{r}
# Add DAPC cluster group information to samples data frame
samples_grp <- data.frame("Sample_ID" = names(grp_meta$grp), "Group" = unname(grp_meta$grp))
# 
samples_keep <- samples %>% 
  filter(Sample_ID %in% samples_grp$Sample_ID) %>% 
  left_join(samples_grp) %>% 
  arrange(Sample_ID)
# 
write_csv(samples_keep, "~/Desktop/samples_keep.csv")
write_csv(samples_grp, "~/Desktop/samples_grp.csv")
```


### DAPC variable contributions
```{r}
contrib <- loadingplot(dapc_meta$var.contr, axis=2,
                       thres=.07, lab.jitter=1)
# 
dapc_loadings <- data.frame(dapc_meta$var.contr) %>% 
  arrange(desc(LD1))
```

### Interpreting group memberships
```{r}
pdf("./outputs/figures_VCF_SNP_assignplot.pdf", width = 8.5, height = 11)
par(mar=c(4,10,4,4))
# summary(dapc_meta)
assignplot(dapc_meta)
dev.off()
```
```{r}
compoplot(dapc_meta, posi="bottomright",
          txt.leg=paste("Cluster", 1:8), lab="",
          ncol=1, xlab="individuals", col=funky(6))
```

