---
title: "Pocillopora meta-tree visualization"
author: "Mike Connelly"
date: "2/9/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

## Setup packages and working directories
```{r}
library("ape")
library("tidyverse")
library("tidytree")
library("treeio")
library("ggtree")
library("ggtreeExtra")
library("ggrepel")
#
library("extrafont")
library("extrafontdb")
```

## Import sample metadata
```{r factor levels}
# study_levels <- c("Mayfield2014", "Vidal-Dupiol2014", "Yuan2017", "Zhou2017", "Brener-Raffali2018", "Delgadillo-Nuno2018", "Tang2018", "Wecker2018", "Zhou2018", "Poquita-Du2019", "Zhou2019", "Buitrago-Lopez2020", "Connelly2020", "Chuang-Mitarai2020", "Li2020", "Connelly2022", "Grupstra2022")
study_levels <- c("Mayfield2014", "Vidal-Dupiol2014", "Yuan2017", "Zhou2017", "Brener-Raffali2018", "Delgadillo-Nuno2018", "Tang2018", "Wecker2018", "Zhou2018", "Poquita-Du2019", "Zhou2019", "Buitrago-Lopez2020", "Chuang-Mitarai2020", "Li2020", "ConnellyEAPSI", "Grupstra2022")
```
```{r}
# Import sample metadata 
samples <- read_csv("data/SampleData_MetaPocillopora_012722.csv") %>% 
  dplyr::filter(Include_SNPs == TRUE) %>% 
  dplyr::mutate("Sample_ID" = str_c(`Study`, `Sample`, sep = "_"))
# 
```

```{r}
#pdam_genome <- ape::read.dna("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.fasta", format = "fasta")
```
```{r}
#pdam_gff <- read.table("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.gff", sep="\t", quote="")
```
```{r colors}
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
gcolors <- gg_color_hue(n_distinct(samples$Study))
```

## Import tree
```{r}
# file <- 
# tree <- read.newick("./outputs/raxml/raxml_run_001/metapocillopora_004_filtered_strict.raxml.support")
# tree <- read.newick("./outputs/phylotrans_pdam/metapocillopora_004_filtered_strict.min4.phy.treefile")
# tree <- read.newick("./outputs/iqtree/iqtree_run_002/metapocillopora_004_filtered_strict.treefile")
tree <- read.newick("./outputs/iqtree/iqtree_run_003/metapocillopora_004_filtered_strict.contree")
# tree <- read.newick("./outputs/iqtree/iqtree_run_004/metapocillopora_004_filtered_strict.treefile")

# inspect phylogenetic tree
tree$tip.label <- str_replace(tree$tip.label, "_1$", "")
```

```{r}
ggtree(tree)
```

```{r}
tree$tip.label %in% samples$Sample_ID
# 
samples <- samples %>% dplyr::filter(Sample_ID %in% tree$tip.label)
df_tip_data <- samples[order(match(tree$tip.label,samples$Sample_ID)), ] %>% 
  relocate(Sample_ID, .before = Study)
df_tip_data$Study <- factor(df_tip_data$Study, levels = study_levels, ordered = TRUE)
```
```{r}
ggtree(tree, layout = "circular", branch.length = "none") %<+% df_tip_data + 
  geom_treescale(x = 0.1, width = 0.05) +
  geom_text(aes(label=node), hjust=-.3)
  # geom_nodelab(aes(label=tree$node.label)) + 
  # geom_tiplab(size = 3, hjust = -0.05) +
  geom_tippoint(aes(fill = Study, shape = Species), size = 3, alpha = 0.8, stroke = 0.5, color = "black", show.legend = TRUE) +
  scale_fill_manual(values = gg_color_hue(n_distinct(samples$Study))) +
  scale_shape_manual(values = c(21, 24, 22, 23)) 
```


```{r}
# Figure 6 
pdf("./outputs/figures/Fig6_tree.pdf", height = 12, width = 13)
tree_title <- expression(paste("Maximum likelihood phylotranscriptomic tree"))
gtree <- ggtree(tree, layout = "rectangular", branch.length = "none") #
# adjust bootstrap values
d <- gtree$data
d <- d[!d$isTip,]
d$label <- as.numeric(d$label)
#
boots <- d #[d$label > 80,]
boots$label <-  ifelse(boots$label == 100, "***",
                       ifelse(boots$label >= 90, "**",
                              ifelse(boots$label >= 70, "*",
                                     ifelse(boots$label >= 50, ".", "")))) 

# plot tree, with labeled clades
# 233 - P. verrucosa; 277 - P. damicornis; 342 - P. acuta
  # geom_hilight(node = 233, fill="steelblue", alpha=.3) +
  # geom_hilight(node = 277, fill="yellow", alpha=.3) +
  # geom_hilight(node = 342, fill="lightgreen", alpha=.3) +
# genotype clades
# 261 - HW1; 291, 274, 282, 302
gtree %<+% df_tip_data + 
  geom_treescale(x = 0, y = 200, width = 0.01, offset = 3) +
  geom_rootedge(rootedge = 0.02) +
  geom_hilight(node = 240, fill="steelblue", alpha=.6) +
  geom_hilight(node = 245, fill="steelblue", alpha=.6) +
  geom_hilight(node = 250, fill="steelblue", alpha=.6) +
  geom_hilight(node = 255, fill="steelblue", alpha=.6) +
  geom_hilight(node = 261, fill="steelblue", alpha=.6) +
  geom_hilight(node = 274, fill="steelblue", alpha=.6) +
  geom_hilight(node = 282, fill="steelblue", alpha=.6) +
  geom_hilight(node = 287, fill="steelblue", alpha=.6) +
  geom_hilight(node = 291, fill="steelblue", alpha=.6) +
  geom_hilight(node = 302, fill="steelblue", alpha=.6) +
  geom_hilight(node = 308, fill="steelblue", alpha=.6) +
  geom_hilight(node = 333, fill="steelblue", alpha=.6) +
  geom_hilight(node = 345, fill="steelblue", alpha=.6) +
  geom_hilight(node = 352, fill="steelblue", alpha=.6) +
  geom_hilight(node = 360, fill="steelblue", alpha=.6) +
  geom_hilight(node = 371, fill="steelblue", alpha=.6) +
  geom_hilight(node = 382, fill="steelblue", alpha=.6) +
  geom_hilight(node = 394, fill="steelblue", alpha=.6) +
  geom_hilight(node = 400, fill="steelblue", alpha=.6) +
  geom_hilight(node = 422, fill="steelblue", alpha=.6) +
  # geom_tiplab(size = 3, hjust = -0.05) +
  geom_tippoint(aes(fill = Study, shape = Species), size = 6, alpha = 0.8, stroke = 0.5, color = "black", show.legend = T) +
  geom_text(data = boots, aes(label=label), size = 2, hjust = 1.4, vjust = -0.4, fontface = 2) +
  # geom_nodelab(aes(label = node), size = 2) +
  scale_fill_manual(values = gg_color_hue(n_distinct(samples$Study)),
                    guide = guide_legend(override.aes = list(size = 6, shape = 21),
                                         nrow = 8, title.position = "top")) +
  scale_shape_manual(values = c(21, 24, 22, 23),
                     guide = guide_legend(override.aes = list(size = 6,
                                                              alpha = 0.8, stroke = 0.5, fill = "black"),
                                          nrow = 4, title.position = "top")) +
  theme(plot.title = element_text(size = 28, hjust = 0.05, vjust = 0),
        legend.title = element_text(size = 24),
        legend.text = element_text(size = 20),
        legend.position = "bottom",
        legend.spacing.y = unit(0.15, "in"),
        plot.margin = margin(t = 0, b = 0, r = 0, l = 0, unit = "in")) +
  ggtitle(tree_title)
dev.off()
```
```{r}
# Figure 6, circular
pdf("./outputs/figures/Fig6_tree_circular.pdf", height = 16, width = 13)
tree_title <- expression(paste("Maximum likelihood phylotranscriptomic tree cladogram (7,688 SNPs)"))
gtree <- ggtree(tree, layout = "circular", branch.length = "none") #
# adjust bootstrap values
d <- gtree$data
d <- d[!d$isTip,]
d$label <- as.numeric(d$label)
#
boots <- d # [d$label == 100,]
boots$label <-  ifelse(boots$label == 100, "***",
                       ifelse(boots$label >= 90, "**",
                              ifelse(boots$label >= 70, "*",
                                     ifelse(boots$label >= 50, "\U00B7", "")))) 

#plot tree
gtree %<+% df_tip_data + 
  # geom_treescale(x = 0.02, y = 0, offset = 3, width = 0.01) +
  geom_rootedge(rootedge = 0.02) +
  # geom_tiplab(size = 3, hjust = -0.05) +
  #
  geom_hilight(node = 240, fill = gcolors[4], alpha = 0.3) +
  geom_hilight(node = 245, fill = gcolors[4], alpha = 0.3) +
  geom_hilight(node = 250, fill = gcolors[4], alpha = 0.3) +
  geom_hilight(node = 255, fill = gcolors[1], alpha = 0.3) +
  geom_hilight(node = 261, fill = gcolors[14], alpha = 0.3) +
  geom_hilight(node = 274, fill = gcolors[5], alpha = 0.3) +
  geom_hilight(node = 282, fill = gcolors[4], alpha = 0.3) +
  geom_hilight(node = 287, fill = gcolors[2], alpha = 0.3) +
  geom_hilight(node = 291, fill = gcolors[3], alpha = 0.3) +
  geom_hilight(node = 302, fill = gcolors[8], alpha = 0.3) +
  geom_hilight(node = 308, fill = gcolors[6], alpha = 0.3) +
  geom_hilight(node = 333, fill = gcolors[7], alpha = 0.3) +
  geom_hilight(node = 345, fill = gcolors[12], alpha = 0.3) +
  geom_hilight(node = 352, fill = gcolors[9], alpha = 0.3) +
  geom_hilight(node = 360, fill = gcolors[9], alpha = 0.3) +
  geom_hilight(node = 371, fill = gcolors[9], alpha = 0.3) +
  geom_hilight(node = 382, fill = gcolors[14], alpha = 0.3) +
  geom_hilight(node = 394, fill = gcolors[12], alpha = 0.3) +
  geom_hilight(node = 400, fill = gcolors[12], alpha = 0.3) +
  geom_hilight(node = 422, fill = gcolors[14], alpha = 0.3) +
  #
  geom_tippoint(aes(fill = Study, shape = Species), size = 6, alpha = 0.8, stroke = 0.5, color = "black", show.legend = T) +
  geom_text(data = boots, aes(label = label, angle = angle), size = 5, hjust = 1.15, vjust = 0.15, fontface = 2) +
  scale_fill_manual(values = gg_color_hue(n_distinct(samples$Study)),
                    guide = guide_legend(override.aes = list(size = 6, shape = 21),
                                         nrow = 8, title.position = "top")) +
  scale_shape_manual(values = c(21, 24, 22, 23),
                     guide = guide_legend(override.aes = list(size = 6,
                                                              alpha = 0.8, stroke = 0.5, fill = "black"),
                                          nrow = 4, title.position = "top")) +
  theme(plot.title = element_text(size = 28, hjust = 0.05),
        # plot.margin = margin(t = 5, b = 5, r = 1, l = 1, unit = "pt"),
        legend.title = element_text(size = 24),
        legend.text = element_text(size = 20),
        legend.position = "bottom",
        legend.spacing.y = unit(0.15, "in")) +
  ggtitle(tree_title)
dev.off()
```
```


