---
title: "Pocillopora transcriptome meta-analysis QC"
author: "Mike Connelly"
date: "1/11/2022"
output: 
  html_document:
      code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
options(stringsAsFactors = FALSE)
```

## Setup packages and working directories
```{r packages}
library("tidyverse")
library("ggrepel")
library("scales")
library("patchwork")
```

## Import sample metadata
```{r factor levels}
study_levels <- c("Mayfield2014", "Vidal-Dupiol2014", "Yuan2017", "Zhou2017", "Brener-Raffali2018", "Delgadillo-Nuno2018", "Tang2018", "Wecker2018", "Zhou2018", "Poquita-Du2019", "Zhou2019", "Buitrago-Lopez2020", "Chuang-Mitarai2020", "Li2020", "Becker2021", "Rutgers2021", "Stephens2021", "Zhu2021", "ConnellyEAPSI", "Grupstra2022")
```
```{r sample_data, echo=TRUE}
# Import sample metadata 
samples <- read_csv("data/SampleData_MetaPocillopora_AllBioProjects.csv") 
```

## Import and tidy transcriptome pipeline QC data
```{r QC_data}
QC_summary_data <- read_csv("./data/MultiQCs_MetaPocillopora_AllBioProjects.csv") #%>%
sum(QC_summary_data$`Raw`)
# 8.56 billion reads as of 07/07/2022
range(QC_summary_data$`Raw`)
mean(QC_summary_data$`Raw`)
sd(QC_summary_data$`Raw`)
```

```{r tidy_QC_data}
QC_summary_select <- QC_summary_data %>% 
  dplyr::select(-`Study`, -`BioProject`) %>% 
  full_join(samples, by = "Sample") %>%
  filter(`Include_Counts` == TRUE) # remove Mayfield samples

QC_summary <- QC_summary_select %>%
  pivot_longer(., `Raw`:`Assigned Reads`, names_to = "pipeline_step", values_to = "read_count") 
#
QC_summary$Study <- factor(QC_summary$Study, levels = study_levels, ordered = TRUE)
#
QC_summary$pipeline_step <- factor(QC_summary$pipeline_step, levels = c("Raw", "Trimmed", "Aligned", "Input Reads", "Assigned Reads"), ordered = TRUE)
#
QC_summary <- QC_summary %>% 
  arrange(`Study`)
```

```{r}
# find sample with most aligned reads for each genotype in each study
 top_aligned_samples <- QC_summary_select %>%
  group_by(Study, Genotype) %>% 
  slice_max(Aligned) %>%
  arrange(desc(Aligned)) %>% 
  mutate(Sample_ID = str_c(Study, Sample, sep = "_"))
#
 top_aligned_samples <- QC_summary_select %>%
   filter(Treatment == "control") %>% 
  group_by(Study, Genotype) %>% 
  slice_max(Aligned) %>%
  arrange(desc(Aligned)) %>% 
  mutate(Sample_ID = str_c(Study, Sample, sep = "_"))
# 32 samples 
top_samples <- top_aligned_samples$Sample_ID
```

## Exploratory plots of sample read depth
```{r colors}
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
gcolors <- gg_color_hue(n_distinct(QC_summary$Study))
```
```{r shapes}
spshapes <- c(21, 24, 22, 23, 23, 25)
```

```{r}
FASTQCg <- QC_summary %>% 
  filter(pipeline_step == "Raw") %>%
  ggplot(aes(Sample, read_count)) +
  geom_hline(yintercept = 10000000, color  = "green") +
  geom_hline(yintercept = 8000000) + 
  geom_hline(yintercept = 1000000, color = "red")
FASTQCg + 
  geom_point(aes(fill = Study, shape = Species), color = "black") + 
  scale_fill_manual(values = gcolors) +
  scale_shape_manual(values = spshapes) +
  theme(axis.text.x = element_text(angle = 90))
  #geom_text_repel(data = FASTQC_summary[1:15, ], aes(label = GNomExID), color = "black", size = 3, box.padding = unit(0.5, "lines")
```

## Figure 2: Read depth and proportions across pipeline steps
```{r read_counts}
pipeQC_counts <- QC_summary %>%
  dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, read_count, color = Study)) +
  geom_path(aes(group = Sample), show.legend = FALSE) +
  geom_point(aes(fill = Study, shape = Species), size = 5, color = "black", alpha = 0.8, stroke = 0.5, show.legend = FALSE) + 
  # facet_grid(.~Study, scales = "free") +
  scale_y_continuous(limits = c(0, 1.25e8)) +
  scale_color_manual(values = gcolors) +
  scale_shape_manual(values = spshapes) +
  scale_fill_manual(values = gcolors) +
  theme_bw() +
    theme(plot.title = element_text(size = 24),
        axis.title = element_text(size = 24),
        axis.text = element_text(size = 20),
        axis.text.x = element_text(angle = 315, hjust = 0),
        # panel.grid.major = element_line(color = "grey92"),
        legend.title = element_text(size = 24),
        legend.text = element_text(size = 20),
        legend.spacing.y = unit(0.15, "in")) +
  ylab("read count") +
  xlab("pipeline step") + 
  ggtitle("read counts")
pipeQC_counts
```

```{r proportions}
QC_summary <- QC_summary %>% 
  dplyr::group_by(Sample) %>%
  mutate(proportion = read_count / read_count[pipeline_step == "Raw"],
         proportion_trimmed = read_count / read_count[pipeline_step == "Trimmed"])
```
```{r trimming_stats}
QC_summary %>% 
  group_by(pipeline_step) %>% 
  # filter(pipeline_step == "Trimmed") %>% 
  summarise(mean = mean(proportion, na.rm = T), 
            sd = sd(proportion, na.rm = T))
# 
range(QC_summary$proportion[QC_summary$pipeline_step == "Trimmed"], na.rm = T)
# 
hist(QC_summary$proportion[QC_summary$pipeline_step == "Trimmed"], breaks = seq(0,1,0.01))
```
```{r alignment_stats}
QC_summary %>% 
  group_by(pipeline_step) %>% 
  # filter(pipeline_step == "Trimmed") %>% 
  summarise(mean = mean(proportion_trimmed, na.rm = T), 
            sd = sd(proportion_trimmed, na.rm = T))
# 
range(QC_summary$proportion_trimmed[QC_summary$pipeline_step == "Aligned"], na.rm = T)
# 
hist(QC_summary$proportion_trimmed[QC_summary$pipeline_step == "Aligned"], breaks = seq(0,1,0.01))
#
QC_summary$proportion_trimmed[QC_summary$pipeline_step == "Aligned"] < 0.25
```

```{r}
pipeQC_props <- QC_summary %>%
  dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, proportion, color = Study, fill = Study)) +
  geom_path(aes(group = Sample), show.legend = TRUE) +
  geom_point(aes(fill = Study, shape = Species), size = 5, alpha = 0.8, stroke = 0.5, color = "black", show.legend = TRUE) + 
  # facet_grid(.~Study, scales = "free") +
  scale_y_continuous(limits = c(0, 1), labels = scales::percent) +
  scale_color_manual(values = gcolors) +
  scale_shape_manual(values = spshapes) +
  scale_fill_manual(values = gcolors) +
  guides(fill = guide_legend(order = 1, override.aes = list(fill = gcolors, shape = 21, stroke = 0.5)),
         shape = guide_legend(order = 2, override.aes = list(shape = spshapes, alpha = 0.8, stroke = 0.5, fill = "black")),
         color = "none") +
  theme_bw() +
  theme(plot.title = element_text(size = 24),
        axis.title = element_text(size = 24),
        axis.text = element_text(size = 20),
        axis.text.x = element_text(angle = 315, hjust = 0),
        # panel.grid.major = element_line(color = "grey92"),
        legend.title = element_text(size = 24),
        legend.text = element_text(size = 20),
        legend.spacing.y = unit(0.15, "in")) +
  ylab("proportion") +
  xlab("pipeline step") + 
  ggtitle("read proportions")
pipeQC_props
```
```{r fig2}
qcfig <- (pipeQC_counts | pipeQC_props) + 
  plot_annotation(tag_levels = "A",
                  title = "RNAseq read depth, trimming, and alignment rates") &
  theme(plot.tag = element_text(size = 28),
        plot.title = element_text(size = 28))
# qcfig + ggtitle("RNAseq read depth, trimming, and alignment rates")
#   theme(plot.title = element_text(size = 24))
ggsave(qcfig, filename = "./outputs/figures/Fig2_pipelineQC_all.pdf", width = 13, height = 8.5, units = "in", device = "pdf")
```

```{r fig2}
qcfig <- (pipeQC_counts | pipeQC_props) + 
  plot_annotation(tag_levels = "A",
                  title = "RNAseq read depth, trimming, and alignment rates") &
  theme(plot.tag = element_text(size = 14),
        plot.title = element_text(size = 14))
# qcfig + ggtitle("RNAseq read depth, trimming, and alignment rates")
#   theme(plot.title = element_text(size = 24))
ggsave(qcfig, filename = "./outputs/figures/Fig2_pipelineQC_all.pdf", width = 6.5, height = 4.25, units = "in", device = "pdf")
```