---
title: "Pocillopora meta-analysis VCF SNP filtering w/ SNPFILTR"
author: "Mike Connelly"
date: "4/26/2022"
output: html_document
---

```{r setup, include=FALSE}
# O'Leary et al. 
# modified from https://github.com/sjoleary/SNPFILT/blob/master/filteringsnapper.Rmd
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())

```
```{r}
library("tidyverse")
library("vcfR")
library("SNPfiltR")
# 
source("./R/VCFfilterstats_edit.R")
source("./R/VCFfilterviz.R")
# need to build custom functions in the repository (ex. multiplot())
```

## Preliminary filtered dataset stats
The raw VCF was filtered on Hydra for loci with <25% missing data, Q > 30, depth 20 - 300, and minor allele frequency > 0.01
The preliminary VCF contains 720,086 sites in 217 individuals.
Target several thousand (~1,000 - 10,000) SNPs for downstream analyses
## Import VCF file
```{r imporf vcf}
# vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/metapocillopora_004_filtered_primary.vcf")
vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/metapocillopora_005_filtered_primary.vcf")

# vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/filtered_vcfs/max-missing-0.75_indv-missing-0.75_max-missing_0.1.recode.vcf")
```
```{r}
### check the metadata present in your vcf
vcf_meta
# Full-cohort VCF
# ***** Object of Class vcfR *****
# 217 samples
# 1306 CHROMs
# 720,086 variants
# Object size: 3310.9 Mb
# 0 percent missing data
# *****        *****         *****
# 
# Genotype-specific VCF
# ***** Object of Class vcfR *****
# 30 samples
# 1430 CHROMs
# 913,935 variants
# Object size: 755 Mb
# 0 percent missing data
# *****        *****         *****

#generate popmap file. Two column popmap with the same format as stacks, and the columns must be named 'id' and 'pop'
popmap <- data.frame(id = colnames(vcf_meta@gt)[2:length(colnames(vcf_meta@gt))],
                     pop = substr(colnames(vcf_meta@gt)[2:length(colnames(vcf_meta@gt))], 1,4))
```
## Inspect with vcfR
```{r}
# pdam_genome <- ape::read.dna("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.fasta", format = "fasta")
# pdam_gff <- read.table("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.gff", sep="\t", quote="")
# 
# pdam_meta_chrom <- create.chromR(name="Supercontig", vcf = vcf_meta, seq = pdam_genome, ann = pdam_gff, verbose = FALSE)
# chrom <- proc.chromR(pdam_meta_chrom, verbose = TRUE)
```


## Filter VCF file
```{r}
vcf_meta %>% SNPfiltR::hard_filter(.)
```
```{r}
#hard filter to minimum depth of 15, and minimum genotype quality of 30
vcf_meta_001 <- hard_filter(vcfR = vcf_meta, depth = 20, gq = 30)
# Genotype-specific VCF
# 44.23% of genotypes fall below a read depth of 20 and were converted to NA
# 1.86% of genotypes fall below a genotype quality of 30 and were converted to NA
# 
vcf_meta_001
```
```{r}
 vcf_meta_002 <- vcf_meta_001 %>%  
  filter_allele_balance() #%>% 
# Genotype-specific VCF
# 41.04% of het genotypes (8.45% of all genotypes) fall outside of 0.25 - 0.75 allele balance ratio and were converted to NA
```
```{r}
max_depth(vcf_meta_002)
# Genotype-specific VCF
# dashed line indicates a mean depth across all SNPs of 73.5
```

```{r}
#run function to visualize samples
missing_by_sample(vcfR = vcf_meta_002, popmap = popmap)
```
```{r}
vcf_meta_003 <- vcf_meta_002 %>% missing_by_sample(., cutoff = .95)
# 2 samples are above a 0.95 missing data cutoff, and were removed from VCF - Grupstra2022
#subset popmap to only include retained individuals
popmap<-popmap[popmap$id %in% colnames(vcf_meta_003@gt),]
#if there are still problematic samples, drop them using the following syntax
#vcfR <- vcfR[,colnames(vcfR@gt) != "A_woodhouseii_24711" & colnames(vcfR@gt) != "A_californica_45901"]

#remove invariant sites generated by dropping individuals
vcf_meta_004 <- min_mac(vcf_meta_003, min.mac = 1)
# Genotype-specific VCF
# 11.63% of SNPs fell below a minor allele count of 1 and were removed from the VCF
```
```{r}
missing_by_snp(vcf_meta_004)
#
vcf_meta_005 <- vcf_meta_004 %>% missing_by_snp(., cutoff = .9)
# 99.63% of SNPs fell below a completeness cutoff of 0.9 and were removed from the VCF
# 98.32% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
# 96.99% of SNPs fell below a completeness cutoff of 0.8 and were removed from the VCF
# 77.7% of SNPs fell below a completeness cutoff of 0.65 and were removed from the VCF
```

## Step 5:
investigate the effect of a minor allele count (MAC) cutoff on downstream inferences.
```{r}
#investigate clustering patterns with and without a minor allele cutoff
#use min.mac() to investigate the effect of multiple cutoffs
vcf_meta_mac <- vcf_meta_005 %>% min_mac(., min.mac = 2)
# 36.14% of SNPs fell below a minor allele count of 2 and were removed from the VCF
```
```{r}
#assess clustering without MAC cutoff
miss <- assess_missing_data_tsne(vcf_meta_005, popmap, clustering = FALSE)

#assess clustering wit MAC cutoff
miss <- assess_missing_data_tsne(vcf_meta_mac, popmap, clustering = FALSE)
```

```{r}
#plot depth per snp and per sample
dp <- extract.gt(vcf_meta_005, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_meta_005, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

```

## Step 6: 
  Write out files for downstream analysis
```{r}

vcf_meta_006 <- distance_thin(vcf_meta_005, min.distance = 1000)
# 2945 out of 13570 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering
# 1173 out of 2995 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering - 0.95
```
```{r}
vcf_meta_thin <- vcf_meta_006
vcf_meta <- vcf_meta_006
```
```{r}
#write out vcf with all SNPs
vcfR::write.vcf(vcf_meta, "~/Downloads/vcf_meta_snapp.vcf.gz")
```
  
  