---
title: "Pocillopora meta-analysis VCF SNP filtering w/ SNPFILTR"
author: "Mike Connelly"
date: "4/26/2022"
output: html_document
---

```{r setup, include=FALSE}
# O'Leary et al. 
# modified from https://github.com/sjoleary/SNPFILT/blob/master/filteringsnapper.Rmd
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())

```
```{r}
library("tidyverse")
library("vcfR")
library("SNPfiltR")
# 
source("./R/VCFfilterstats_edit.R")
source("./R/VCFfilterviz.R")
# need to build custom functions in the repository (ex. multiplot())
```
## Import sample metadata
```{r factor levels}
study_levels <- c("Vidal-Dupiol2013", "Mayfield2014", "Vidal-Dupiol2014", "Yuan2017", "Zhou2017", "Brener-Raffali2018", "Delgadillo-Nuno2018", "RSMAS2018", "Tang2018", "Wecker2018", "Zhou2018", "Poquita-Du2019", "Zhou2019", "Buitrago-Lopez2020", "Chuang-Mitarai2020", "Li2020", "Becker2021", "Rutgers2021", "Stephens2021", "Zhu2021", "ConnellyEAPSI", "Grupstra2022")
```
```{r}
# Import sample metadata 
samples <- read_csv("data/SampleData_MetaPocillopora_AllBioProjects.csv") %>% 
  dplyr::filter(Include_SNPs == TRUE) %>% 
  dplyr::mutate("Sample_ID" = str_c(`Study`, `Sample`, sep = "_"))
# 
# samples %>% filter(Sample_ID %in% popmap$id) # just retained genotype-specific samples
```
## Preliminary filtered dataset stats
The raw VCF was filtered on Hydra for loci with <25% missing data, Q > 30, depth 20 - 300, and minor allele frequency > 0.01

The preliminary full-cohort VCF contains 720,086 sites in 217 individuals.
The preliminary genotype-specific VCF contains 

Target several thousand (~1,000 - 10,000) SNPs for downstream analyses
## Import VCF file
```{r imporf_vcf}
# vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/metapocillopora_004_filtered_primary.vcf")
# vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/metapocillopora_007_filtered_primary.vcf")
# vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/filtered_vcfs/max-missing-0.75_indv-missing-0.75_max-missing_0.1.recode.vcf")
```
```{r imporf_vcf}
# Final samples both genotype-specific VCF
vcf_meta <- vcfR::read.vcfR(file = "./outputs/phylotrans_pdam/metapocillopora_final_samples_both_filtered_primary.vcf")
```

```{r}
### check the metadata present in your vcf
vcf_meta
# Full-cohort VCF
# ***** Object of Class vcfR *****
# 217 samples
# 1306 CHROMs
# 720,086 variants
# Object size: 3310.9 Mb
# 0 percent missing data
# *****        *****         *****
# 
# Genotype-specific VCF
# ***** Object of Class vcfR *****
# 30 samples
# 1430 CHROMs
# 913,935 variants
# Object size: 755 Mb
# 0 percent missing data
# *****        *****         *****
#
# Final samples both genotype-specific VCF
# ***** Object of Class vcfR *****
# 57 samples
# 1522 CHROMs
# 717,733 variants
# Object size: 1253.4 Mb
# 0 percent missing data
# *****        *****         *****
```
```{r popmap}
# generate popmap file. Two column popmap with the same format as stacks, and the columns must be named 'id' and 'pop'
# consider which population to assign samples - study, species ID, other groupings?
popmap <- data.frame(id = colnames(vcf_meta@gt)[2:length(colnames(vcf_meta@gt))],
                     pop = substr(colnames(vcf_meta@gt)[2:length(colnames(vcf_meta@gt))], 1,4))
#
popmap <- popmap %>% 
  dplyr::mutate(Sample_ID = str_replace(`id`, "_1$", "")) %>% 
  left_join(samples, by = "Sample_ID") %>% 
  dplyr::select(id:Species)

#save popmap for SNAPP, other assignments
write_csv(popmap, "./data/popmap_genotypes.csv")
```
## Inspect with vcfR
```{r}
# pdam_genome <- ape::read.dna("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.fasta", format = "fasta")
# pdam_gff <- read.table("/Users/mikeconnelly/computing/sequences/genomes/coral/pocillopora/pdam/pdam_genome.gff", sep="\t", quote="")
# 
# pdam_meta_chrom <- create.chromR(name="Supercontig", vcf = vcf_meta, seq = pdam_genome, ann = pdam_gff, verbose = FALSE)
# chrom <- proc.chromR(pdam_meta_chrom, verbose = TRUE)
```

## Filter VCF file
```{r}
vcf_meta %>% SNPfiltR::hard_filter(.)
```
### Step 1: Minimum Depth and Read Quality
```{r}
#hard filter to minimum depth of 15, and minimum genotype quality of 30
vcf_meta_001 <- hard_filter(vcfR = vcf_meta, depth = 20, gq = 30)
# Genotype-specific VCF
# 44.23% of genotypes fall below a read depth of 20 and were converted to NA
# 1.86% of genotypes fall below a genotype quality of 30 and were converted to NA
# 
# Final samples both genotype-specific VCF
# 35.68% of genotypes fall below a read depth of 20 and were converted to NA
# 2.3% of genotypes fall below a genotype quality of 30 and were converted to NA
vcf_meta_001
```
### Step 2: Allele Balance
```{r}
 vcf_meta_002 <- vcf_meta_001 %>%  
  filter_allele_balance() #%>% 
# Genotype-specific VCF
# 41.04% of het genotypes (8.45% of all genotypes) fall outside of 0.25 - 0.75 allele balance ratio and were converted to NA
#
# Final samples both genotype-specific VCF
# 31.09% of het genotypes (6.81% of all genotypes) fall outside of 0.25 - 0.75 allele balance ratio and were converted to NA
vcf_meta_002
```
```{r}
max_depth(vcf_meta_002)
# Genotype-specific VCF
# dashed line indicates a mean depth across all SNPs of 73.5
# 
# Final samples both genotype-specific VCF
# dashed line indicates a mean depth across all SNPs of 69.7
```

```{r}
#run function to visualize samples
missing_by_sample(vcfR = vcf_meta_002, popmap = popmap)
```
### Step 3: Apply missing data cutoff by sample
```{r}
vcf_meta_003 <- vcf_meta_002 %>% missing_by_sample(., cutoff = .85) #0.95
# 2 samples are above a 0.95 missing data cutoff, and were removed from VCF - Grupstra2022
#
# Final samples both genotype-specific VCF
# 4 samples are above a 0.95 missing data cutoff, and were removed from VCF - all Connelly EAPSI
# 7 samples are above a 0.85 missing data cutoff, and were removed from VCF
# 13 samples are above a 0.6 missing data cutoff, and were removed from VCF

#subset popmap to only include retained individuals
popmap <- popmap[popmap$id %in% colnames(vcf_meta_003@gt),]
#if there are still problematic samples, drop them using the following syntax
#vcfR <- vcfR[,colnames(vcfR@gt) != "A_woodhouseii_24711" & colnames(vcfR@gt) != "A_californica_45901"]
```

### Step 4: Remove invariant sites generated by dropping individuals
```{r}
#remove invariant sites generated by dropping individuals
vcf_meta_004 <- min_mac(vcf_meta_003, min.mac = 1)
# 
vcf_meta_004
# Genotype-specific VCF
# 11.63% of SNPs fell below a minor allele count of 1 and were removed from the VCF
# 
# Final samples both genotype-specific VCF
# 3.37% of SNPs fell below a minor allele count of 1 and were removed from the VCF
# ***** Object of Class vcfR *****
# 53 samples
# 1513 CHROMs
# 693,513 variants
# Object size: 936.4 Mb
# 37.02 percent missing data
# *****        *****         *****
```

### Step 5: Apply missing data cutoff by SNP
```{r}
missing_by_snp(vcf_meta_004)
#
vcf_meta_005 <- vcf_meta_004 %>% missing_by_snp(., cutoff = 1)
# note, SNAPP requires that all taxa/species have data at each position, so for all-samples are unique species models, cutoff = 1

# Genotype-specific VCF
# 99.63% of SNPs fell below a completeness cutoff of 0.9 and were removed from the VCF
# 98.32% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF
# 96.99% of SNPs fell below a completeness cutoff of 0.8 and were removed from the VCF
# 77.7% of SNPs fell below a completeness cutoff of 0.65 and were removed from the VCF
# 
# Final samples both genotype-specific VCF
# 99.59% of SNPs fell below a completeness cutoff of 0.95 and were removed from the VCF
# ***** Object of Class vcfR *****
# 53 samples
# 481 CHROMs
# 2,841 variants
# Object size: 6.2 Mb
# 3.128 percent missing data
# *****        *****         *****
# 94.4% of SNPs fell below a completeness cutoff of 0.9 and were removed from the VCF
# ***** Object of Class vcfR *****
# 53 samples
# 795 CHROMs
# 38,843 variants
# Object size: 72.6 Mb
# 7.531 percent missing data
# *****        *****         *****
# 88.61% of SNPs fell below a completeness cutoff of 0.85 and were removed from the VCF

vcf_meta_005
```
```{r}
missing_by_sample(vcfR = vcf_meta_005, popmap = popmap)
# 
missing_by_snp(vcf_meta_005)
```

### Step 6: Investigate the effect of a minor allele count (MAC) cutoff on downstream inferences.
```{r}
#investigate clustering patterns with and without a minor allele cutoff
#use min.mac() to investigate the effect of multiple cutoffs
vcf_meta_mac <- vcf_meta_005 %>% min_mac(., min.mac = 2)
# Genotype-specific VCF
# 36.14% of SNPs fell below a minor allele count of 2 and were removed from the VCF
#
# Final samples both genotype-specific VCF
# 0% of SNPs fell below a minor allele count of 2 and were removed from the VCF

```
```{r}
#assess clustering without MAC cutoff
miss <- assess_missing_data_tsne(vcf_meta_005, popmap, clustering = FALSE)

#assess clustering wit MAC cutoff
miss <- assess_missing_data_tsne(vcf_meta_mac, popmap, clustering = FALSE)
```

```{r}

#plot depth per snp and per sample
dp <- extract.gt(vcf_meta_005, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_meta_005, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)
```

### Step 7: Write out files for downstream analysis
```{r}

vcf_meta_006 <- distance_thin(vcf_meta_005, min.distance = 1000)
# Genotype-specific VCF
# 2945 out of 13570 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering
# 1173 out of 2995 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering - 0.95

#
# Final samples both genotype-specific VCF
# 0.95 - 1131 out of 2841 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering
# 0.9 - 6260 out of 38843 input SNPs were not located within 1000 base-pairs of another SNP and were retained despite filtering
vcf_meta_006
```
```{r}
vcf_meta_thin <- vcf_meta_006

# for vcf analysis
vcf_meta <- vcf_meta_005 # 0.9 linked, 38843 SNPs
                         # 0.95 linked, 2841 SNPs
#
vcf_meta <- vcf_meta_006 # 0.9 unlinked, 6260 SNPs
                         # 0.95 unlinked, 1131 SNPs
```
  
## Final VCF summary statistics and figures
### 95% complete... linked
```{r}
vcf_meta_0.95_linked <- vcf_meta_005[is.biallelic(vcf_meta_005), ] 
vcf_meta_0.95_linked
# ***** Object of Class vcfR *****
# 53 samples
# 481 CHROMs
# 2,834 variants
# Object size: 6.2 Mb
# 3.129 percent missing data
# *****        *****         *****
max_depth(vcf_meta_0.95_linked)
# mean depth across all SNPs of 96.8
missing_by_sample(vcfR = vcf_meta_0.95_linked, popmap = popmap)

#plot depth per snp and per sample
dp <- extract.gt(vcf_meta_0.95_linked, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_meta_0.95_linked, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

#write out vcf with all SNPs
vcfR::write.vcf(vcf_meta_0.95_linked, "./outputs/vcfs/vcf_meta_snapp_0.95_linked.vcf.gz")
```

### 95% complete... thinned, unlinked
Ran SNAPP on this 1,131 x 53 sample matrix --> 1101 SNPs dropped in all-taxa model
```{r}
vcf_meta_0.95_unlinked <- vcf_meta_006[is.biallelic(vcf_meta_006), ] 
vcf_meta_0.95_unlinked
# ***** Object of Class vcfR *****
# 53 samples
# 481 CHROMs
# 1,130 variants
# Object size: 3 Mb
# 3.229 percent missing data
# *****        *****         *****
max_depth(vcf_meta_0.95_unlinked)
# mean depth across all SNPs of 92.4
missing_by_sample(vcfR = vcf_meta_0.95_unlinked, popmap = popmap)

pdf("./outputs/figures/vcf_DP_GQ_heatmaps.pdf", width = 6.5, height = 4)
#plot depth per snp and per sample
dp <- extract.gt(vcf_meta_0.95_unlinked, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_meta_0.95_unlinked, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)
dev.off()

#write out vcf with all SNPs
vcfR::write.vcf(vcf_meta_0.95_unlinked, "./outputs/vcfs/vcf_meta_snapp_0.95_unlinked.vcf.gz")
```

Now, convert vcfs to phylip and NEXUS format with vcf2phylip.py scripts, and use in RAxML and SNAPP phylogenetic analysis. Also use for PCA and clustering analyses to discover patterns within data and create discovery-based species hypotheses.

### 95% complete... thinned, unlinked
Ran SNAPP on this 1,131 x 53 sample matrix
```{r}
vcf_meta_1_unlinked <- vcf_meta_006[is.biallelic(vcf_meta_006), ] 
vcf_meta_1_unlinked
# ***** Object of Class vcfR *****
# 53 samples
# 70 CHROMs
# 76 variants
# Object size: 0.7 Mb
# 0 percent missing data
# *****        *****         *****
max_depth(vcf_meta_1_unlinked)
# mean depth across all SNPs of 95.8
missing_by_sample(vcfR = vcf_meta_1_unlinked, popmap = popmap)

#plot depth per snp and per sample
dp <- extract.gt(vcf_meta_1_unlinked, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_meta_1_unlinked, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

#write out vcf with all SNPs
vcfR::write.vcf(vcf_meta_1_unlinked, "./outputs/vcfs/vcf_meta_snapp_1_unlinked.vcf.gz")
```

### 85% complete samples, 100% complete SNPs... thinned, unlinked
Ran SNAPP on this 1,198 x 50 sample matrix
```{r}
vcf_meta_2_unlinked <- vcf_meta_006[is.biallelic(vcf_meta_006), ] 
vcf_meta_2_unlinked
# ***** Object of Class vcfR *****
# 50 samples
# 482 CHROMs
# 1,198 variants
# Object size: 3.1 Mb
# 0 percent missing data
# *****        *****         *****
max_depth(vcf_meta_2_unlinked)
# mean depth across all SNPs of 85.6
missing_by_sample(vcfR = vcf_meta_2_unlinked, popmap = popmap)

#plot depth per snp and per sample
dp <- extract.gt(vcf_meta_2_unlinked, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf_meta_2_unlinked, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

#write out vcf with all SNPs
vcfR::write.vcf(vcf_meta_2_unlinked, "./outputs/vcfs/vcf_meta_snapp_2_unlinked.vcf.gz")
```