---
title: "Pocillopora transcriptome meta-analysis QC"
author: "Mike Connelly"
date: "1/11/2022"
output: 
  html_document:
      code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
options(stringsAsFactors = FALSE)
```
```{r packages}
library("tidyverse")
library("ggrepel")
library("scales")
library("patchwork")
```
```{r data}
# samples <- 
# samples$SampleID <- gsub('-','.', samples$SampleID)
QC_summary <- read_csv("./data/MultiQCs_MetaPocillopora_011122.csv")
sum(QC_summary$`Raw Reads`)
# 1.43 billion reads as of 01/11/2022
```
```{r tidydata}
QC_summary <- QC_summary %>%
  pivot_longer(., `Raw Reads`:`Assigned Reads`, names_to = "pipeline_step", values_to = "read_count") #%>%
  # mutate(partner = ifelse(str_detect(pipeline_step, "symbiont"), "symbiont", "coral")) %>% 
  # left_join(samples[ ,1:6], ., by = "SampleID")
# QC_summary$Treatment <- factor(QC_summary$Treatment, levels = c("control", "Heat", "Antibiotics", "Antibiotics.Heat"), ordered = TRUE)
#
QC_summary$pipeline_step <- factor(QC_summary$pipeline_step, levels = c("Raw Reads", "Trimmed Reads", "Uniquely Mapped Reads", "Input Reads", "Assigned Reads"), ordered = TRUE)
```


### #1-#15 below 8 million reads
### #16-#43 below 10 million reads

```{r}
FASTQCg <- QC_summary %>% 
  filter(pipeline_step == "Raw Reads") %>%
  ggplot(aes(Sample, read_count, color = Study))
FASTQCg + geom_point() + 
  # scale_color_manual(values = condcolors_AxH) + 
  # scale_shape_manual(values = colshapes) +
  geom_hline(yintercept = 10000000, color  = "green") +
  geom_hline(yintercept = 8000000) + 
  geom_hline(yintercept = 1000000, color = "red") +
  theme(axis.text.x = element_text(angle = 90))
  #geom_text_repel(data = FASTQC_summary[1:15, ], aes(label = GNomExID), color = "black", size = 3, box.padding = unit(0.5, "lines")
```

```{r read_counts}
pipeQC_counts <- QC_summary %>%
  dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, read_count, color = Study)) +
  geom_path(aes(group = Sample), show.legend = FALSE) +
  geom_point(aes(fill = Study), size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = TRUE) + 
  facet_grid(.~Study, scales = "free") +
  # scale_y_continuous(limits = c(0, 2.5e7), breaks = c(0, 5e6, 1e7, 1.5e7, 2e7, 2.5e7)) +
  # scale_color_manual(values = condcolors_AxH) +
  # scale_shape_manual(values = colshapes) + 
  # scale_fill_manual(values = condcolors_AxH) +
  theme(axis.text.x = element_text(angle = 330), panel.grid.major = element_line(color = "grey92")) +
  ylab("read count") +
  xlab("pipeline step") + 
  ggtitle("read counts")
pipeQC_counts
```

```{r proportions}
QC_summary <- QC_summary %>% 
  dplyr::group_by(Sample) %>%
  mutate(proportion = read_count / read_count[pipeline_step == "Raw Reads"])
```
```{r}
pipeQC_props <- QC_summary %>%
  dplyr::filter(pipeline_step != "Input Reads" & pipeline_step != "Assigned Reads") %>%
  ggplot(aes(pipeline_step, proportion, color = Study, fill = Study)) +
  geom_path(aes(group = Sample), show.legend = TRUE) +
  geom_point(size = 2, shape = 21, color = "black", alpha = 0.5, stroke = 0.5, show.legend = TRUE) + 
  facet_grid(.~Study, scales = "free") +
  # scale_y_continuous(limits = c(0, 1)) +
  # scale_color_manual(values = condcolors_AxH) +
  # scale_shape_manual(values = colshapes) + 
  # scale_fill_manual(values = condcolors_AxH) +
  theme(axis.text.x = element_text(angle = 330), panel.grid.major = element_line(color = "grey92")) +
  # guides(color = guide_legend(override.aes = list(color = condcolors_AxH, alpha = 1, stroke = 1)),
  #          fill = guide_legend(override.aes = list(fill = condcolors_AxH, shape = 21, alpha = 1, stroke = 0.5)),
  #          shape = guide_legend(override.aes = list(shape = colshapes, alpha = 1, stroke = 0.5))) +
  ylab("proportion raw reads") +
  xlab("pipeline step") + 
  ggtitle("read proportions")
pipeQC_props
```
```{r}
qcfig <- (pipeQC_counts | pipeQC_props) + plot_annotation(tag_levels = "A")
ggsave(qcfig, filename = "./manuscript_figures/FigS5_readcounts.pdf", width = 220, height = 90, units = "mm", device = "pdf")
```

